rfcs/53-consistent-hierarchy.md

```diff
@@ -0,0 +1,421 @@
+dedicated hierarchy management commands.
+
+`PreviousParent` will also be removed and replaced with `ChildAdded`,
+`ChildRemoved` and `ChildMoved` events instead. These will signal changes in the
+hierachy instead of relying on change detection on `Parent` and `PreviousParent`.
+
+## Implementation strategy
+This design attempts to minimize the changes made to the current design while
+addressing the issues enumerated above. The core changes to be made are as
+follows:
+
+ - Make `Parent` and `Children` (mostly) publicly immutable.
+ - Make mutations to the hierarchy rely solely on commands.
+ - Update hierarchy commands to automatically update both `Parent` and
+   `Children` to remain consistent with each other.
+ - Remove the `parent_update_system`.
+ - Remove `PreviousParent`.
+ - Register several hierarchy events that fires whenever time a new child is
+   added or removed from a parent.
+ - Create a `HiearchyQuery` custom SystemParam for more easily iterating over
+   components in the hierarchy.
+
+This change will repurpose the following existing `EntityCommand` extension
+functions for this purpose:
+
+ - `add_child`
+ - `push_children`
+ - `insert_children`
+ - `remove_child` (this will be added as a wrapper around `remove_children`)
+ - `remove_children`
+
+An additional `set_parent(Option<Entity>)` command should also be added as a
+child-side way of altering the hierarchy, defaulting to adding the entity as the
+last child to the provided parent if provided, and detaching the child from it's
+parent if `None`.
+
+### HierarchyQuery
+An extension to this design is to create dedicated `HierarchyQuery<Q, F>`, which
+is a custom `SystemParam` that wraps multiple `Query` objects to simplify
+querying for items in the hierarchy:
+
+ - `roots(_mut)` - returns an iterator over entities that match the provided
+   query parameters that do not have a parent.
+ - `parent(_mut)` - returns single result from the parent of an entity.
+ - `iter_children(_mut)` - returns an iterator over `(parent, child)` results
+   from a query parameter. This does not return Children of Children.
+ - `iter_descendants(_mut)` - returns an iterator over `(parent, child)` results
+   from a query parameter. This does a depth-first traversal over the hierarchy
+   starting at some entity.
+ - `iter_ancestors(_mut)` - returns an iterator over `(parent, child)` results
+   from a query parameter. This bubbles up queries from the bottom to the top
+   starting at some entity.
+
+As suggested by the generic parameters, the query results and filters are
+identical to the ones used by a normal Query or QueryState.
+
+The main motivation for this is to make creating systems like
+`transform_propagate_system` easier. These queries can be generically used for
+any system that needs some form of hierachy traversal. These queries can be quite
+expensive to run, so the documentation for them should reflect this and
+discourage misuing it.
+
+### HierarchyEvents
```

Comment:
I think its worth pointing out that in practice, this will increase the memory usage of the hierarchy system to be something like `2 * size_of::<Entity>() * MAX_HIERARCHY_CHANGES_ACROSS_TWO_FRAMES`. Given that hierarchical spawning for scenes will be pretty burst-ey, in practice this will be a per-hierarchical-entity `2 * size_of::<Entity>()` price unless we add garbage collection to Events.

Although that should be weighed against the cost of `PreviousParent(Entity)`, which includes both additional memory footprint and (sometimes) additional runtime costs for archetype moves.

Reply:
`PreviousParent` is a `size_of::<Entity>()` memory cost and arbitrary archetype moves. Given that `(Global)Transform` is 48 bytes, and one additional `Entity` is 8, just in memory write costs alone, we'd need to write 6x (16 bytes vs 96) the events to compare an archetype move for your typical transform hierarchy change, and that's not counting the archetype/table move overhead, the other components being moved, and the additional archetype fragmentation from the two-fold moves. It's just we incur the cost of `PreviousParent` one frame later instead of immediately.

Building automatic event garbage collection seems perfectly workable given that events already track readers.